usr/lib/common/tok_spec_struct.h

```diff
@@ -275,6 +275,8 @@ struct token_specific_struct {
                                 CK_BYTE *, CK_ULONG , CK_BYTE *, CK_ULONG *);
 
     CK_RV(*t_set_attribute_values) (STDLL_TokData_t *, OBJECT *, TEMPLATE *);
+
+    CK_RV(*t_set_pkey_attr) (STDLL_TokData_t *, TEMPLATE *);
```

Comment:
The name set_pkey_attr is still quite generic. It refers to pkey, that's great, and that it is somehow supposed to set attributes, but it is not clear from the name when this function is called (i.e. during key object creation - generate, create, wrap, unwrap), and it is not clear if the function is called when common code sets attributes, or to let the token set attributes (the second is the intend of this function) . 

Also, there is also a token_specific_set_attribute_values which kind of sounds similar (but has a totally different meaning). 

And there is another similar function: token_specific_object_add. This one is also called during object creation. What's the difference between your new function and and token_specific_object_add? Why can't you use token_specific_object_add instead of yours? I guess the answer is because token_specific_object_add is only called for C_CreateObjetc, but not for the other object creation cases (Generate, Unwrap, Derive), but yours is called for all those cases. 

I see 2 options: 
- either we make it really generic, so that it is called for ALL objects, and lets the token add (and modify?) attributes that it needs, totally independent of PKEY (but PKEY uses it as first),
- or make it very PKEY specific, i.e. change the name to something that makes it clear that it is called during key object creation. 

Remember: This is common code, so it should be generic, so that it could be used by all tokens.

Reply:
I like option 1 and also already thought about that: introduce a generic routine that allows any token add/modify any attributes, but pkey is just the first one. But there is some danger that such a generic routine would get overloaded. Well, on the other hand we dont' want to introduce one separate routine for separate attributes that might come in future.
Option 2 is more or less what the code currently does, so let me think about a better name and leave it for now.

Reply:
The routine name is now: token_specific_set_attrs_for_new_object.